Sources/SwiftBasicFormat/BasicFormat.swift

```diff
@@ -0,0 +1,278 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+open class BasicFormat: SyntaxRewriter {
+  /// How much indentation should be added at a new indentation level.
+  public let indentationIncrement: Trivia
+
+  /// As we reach a new indendation level, its indentation will be added to the
+  /// stack. As we exit that indentation level, the indendation will be popped.
+  public var indentationStack: [Trivia]
+
+  /// The trivia by which tokens should currently be indented.
+  public var currentIndentationLevel: Trivia {
+    return indentationStack.last!
+  }
+
+  /// For every token that is being put on a new line but did not have
+  /// user-specified indentation, the generated indentation.
+  ///
+  /// This is used as a reference-point to indent user-indented code.
+  var anchorPoints: [TokenSyntax: Trivia] = [:]
+
+  public init(indentationIncrement: Trivia = .spaces(4), initialIndentation: Trivia = []) {
```

Comment:
Not sure if I like the default value is `[]`. I think it's a but hard to understand maybe for new developers.

But on the the other hand I cannot find any other name than `.zero`, `.none`, `.empty`. And I'm not convinced they fit either.

Reply:
We used to have `Trivia.zero` but IIRC @bnbarham removed it because it was redundant with the empty array. Do you have strong options on whether we should add it (or `.empty`) back in again?

---

In line 36: Well, `indentationStack` is an array of trivia again. And arrays of arrays aren’t that uncommon, so I think that should be fine…